File: onderwijs4u-back-end/src/main/java/nl/han/devops/interceptor/TokenInterceptor.java
package nl.han.devops.interceptor;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import nl.han.devops.service.AuthenticationService;
import org.springframework.web.servlet.HandlerInterceptor;

public class TokenInterceptor implements HandlerInterceptor {
    private final AuthenticationService authenticationService;

    public TokenInterceptor(AuthenticationService authenticationService) {
        this.authenticationService = authenticationService;
    }

    @Override
    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception {
        String requestURI = request.getRequestURI();
        if (requestURI.endsWith("/login") || requestURI.endsWith("/signup") || requestURI.endsWith("/image")) {
            return true;
        }

        String token = request.getParameter("token");
        if (token == null || token.isEmpty()) {
            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Token missing");
            return false;
        }

        if (!authenticationService.tokenExists(token)) {
            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Invalid token");
            return false;
        }

        return true;
    }
}

File: back-end/src/main/kotlin/nl/lpp/uncenshow/interceptor/TokenInterceptor.kt
package nl.lpp.uncenshow.interceptor

import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import nl.lpp.uncenshow.service.authentication.AuthenticationService
import org.springframework.web.servlet.HandlerInterceptor

class TokenInterceptor (private val authenticationService: AuthenticationService) : HandlerInterceptor {
    @Throws(Exception::class)
    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {

        val requestURI = request.requestURI
        if (requestURI.endsWith("/login") || requestURI.endsWith("/signup") || requestURI.endsWith("/image")) {
            return true
        }

        val token = request.getParameter("token")
        if (token == null || token.isEmpty()) {
            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Token missing")
            return false
        }

        if (!authenticationService.tokenExists(token)) {
            response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Invalid token")
            return false
        }

        return true
    }
}